### PR TITLE
find cause of genesis contract duplicate in random Websocket testing

### LINKED ISSUE #4992: Robust random testing of websocket endpoints
The current websocket endpoint tests are painstakingly hand-written and cover only simple cases of contract creation/archival. Moreover, they are not robust in the face of conflation or heartbeats, in the way we expect applications to be.

Instead, while generating arbitrary contract creations and exercises would be excessive, we can generate sequences of IOU creates and Iou_Split exercises, then observe a running websocket on them. `ContractDelta` gives us the tools to deterministically interpret the results in the presence of conflation and heartbeats. We can use the offsets from these streams to verify that rescanning gives us the same results from that offset on for each resulting sequence.

We can use a different pair of template and choice to test fetch by key streams, checking the additional invariant that the key->value table is consistent at each step, or use a template and choice that are appropriate for testing both query and fetch-by-key.

This would help us to be more confident in the behavior of websocket streams so as to let us declare them stable.

### PR DESCRIPTION
**Pull request by S11001001**
Direct followup to #5879, searching for the cause of problems uncovered by that testing.

As [described below](#issuecomment-628839195), we found a race condition in sandbox-classic's in-memory ACS. Fixing that fixes the duplicate contract problem, and the random Iou_Split test now succeeds without issues.

Also reenable said random testing. Fixes #4992.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by S11001001** [comment]
At first glance, it appears that the `offset` that terminates the `GetActiveContractsResponse` stream is simply the wrong offset. In the below sample, the response states that the ACS ends at offset 0000000000000002, but the transaction with offset 0000000000000003 clearly includes an element of the ACS.

We observe from `ActiveContractSetClient#getActiveContracts` and `TransactionClient#getTransactions`

```
18.388 [dd13] INFO  LedgerClientJwt$ - s11 gacr: GetActiveContractsResponse(,,Vector(THE_GENESIS_CONTRACT),None)
18.388 [dd13] INFO  LedgerClientJwt$ - s11 gacr: GetActiveContractsResponse(0000000000000002,,Vector(),None)
18.392 [dd16] INFO  LedgerClientJwt$ - s11 tx stream start=LedgerOffset(Absolute(0000000000000002)): Transaction(3,0efae40d-3b80-4956-9a99-5b62c569f2c1,,Some(Timestamp(1589483298,384000000)),Vector(Event(Created(THE_GENESIS_CONTRACT))),0000000000000003,None)

where

THE_GENESIS_CONTRACT = CreatedEvent(#3:0,#3:0,Some(Identifier(d5c0a0e21d368bdb972616c3c84fae08f3fc4da3b5d819d1671edc7b9399a211,Iou,Iou)),None,Some(Record(Some(Identifier(d5c0a0e21d368bdb972616c3c84fae08f3fc4da3b5d819d1671edc7b9399a211,Iou,Iou)),Vector(RecordField(issuer,Some(Value(Party(Alice)))), RecordField(owner,Some(Value(Party(Alice)))), RecordField(currency,Some(Value(Text(USD)))), RecordField(amount,Some(Value(Numeric(52.0000000000)))), RecordField(observers,Some(Value(List(List(Vector())))))))),Vector(Alice),Vector(Alice),Vector(),Some())
```

Given the logging of (on advice of @leo-da )

```diff
diff --git a/ledger-service/http-json/src/main/scala/com/digitalasset/http/LedgerClientJwt.scala b/ledger-service/http-json/src/main/scala/com/digitalasset/http/LedgerClientJwt.scala
index 0533973b75..6b24c79d8d 100644
--- a/ledger-service/http-json/src/main/scala/com/digitalasset/http/LedgerClientJwt.scala
+++ b/ledger-service/http-json/src/main/scala/com/digitalasset/http/LedgerClientJwt.scala
@@ -24,7 +24,7 @@ import scalaz.OneAnd
 
 import scala.concurrent.{ExecutionContext, Future}
 
-object LedgerClientJwt {
+object LedgerClientJwt extends com.typesafe.scalalogging.StrictLogging {
 
   type SubmitAndWaitForTransaction =
     (Jwt, SubmitAndWaitRequest) => Future[SubmitAndWaitForTransactionResponse]
@@ -84,6 +84,7 @@ object LedgerClientJwt {
       client.activeContractSetClient
         .getActiveContracts(filter, verbose, bearer(jwt))
         .mapMaterializedValue(_ => NotUsed)
+      .wireTap{gacr => logger.info(s"s11 gacr: $gacr")}
 
   sealed abstract class Terminates extends Product with Serializable {
     import Terminates._
@@ -102,6 +103,7 @@ object LedgerClientJwt {
   private val ledgerEndOffset =
     LedgerOffset(LedgerOffset.Value.Boundary(LedgerOffset.LedgerBoundary.LEDGER_END))
 
+  @SuppressWarnings(Array("org.wartremover.warts.Product", "org.wartremover.warts.Serializable"))
   def getCreatesAndArchivesSince(client: LedgerClient): GetCreatesAndArchivesSince =
     (jwt, filter, offset, terminates) => {
       val end = terminates.toOffset
@@ -110,6 +112,7 @@ object LedgerClientJwt {
       else
         client.transactionClient
           .getTransactions(offset, terminates.toOffset, filter, verbose = true, token = bearer(jwt))
+          .wireTap(tx => logger.info(s"s11 tx stream start=$offset: $tx"))
     }
 
   private def skipRequest(start: LedgerOffset, end: Option[LedgerOffset]): Boolean = {
```

**Comment by gerolf-da** [comment]
@S11001001 Is that with Sandbox Classic or Sandbox?

**Comment by S11001001** [comment]
@gerolf-da I believe all these tests use classic.

https://github.com/digital-asset/daml/blob/40d40a95cdac0a88fcc5476276cba4ad8f612676/ledger-service/http-json/src/test/scala/com/digitalasset/http/HttpServiceTestFixture.scala#L29

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala [hunk 1]
```diff
@@ -196,45 +196,49 @@ class InMemoryLedger(
...
+                signatories = contract.signatories.toSeq,
+                observers = contract.observers.toSeq,
+                agreementText = Some(contract.agreementText)
+              )
             )
           )
-        )
-      }
+        },
+      ledgerEnd /*TODO SC derive from copy of acs var instead*/ )
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala:241]: What you can do is something like this at the beginning:
```
val (activeContracts, ledgerEnd) = this.synchronized { (acs.activeContracts, ledgerEnd) }
```

This way `ledgerEnd` and `acs` will be "in sync".

Comment (S11001001) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala:241]: With that change, the test no longer finds a duplicate.

Comment (S11001001) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala]: 907a64122e

**Pull request merged**